File: app/controllers/LoginController.java
package controllers;

import controllers.core.BaseController;
import model.StringConsts$;
import model.user.Course;
import model.user.Course$;
import model.user.User;
import play.Environment;
import play.data.DynamicForm;
import play.data.FormFactory;
import play.mvc.Result;

import javax.inject.Inject;

public class LoginController extends BaseController {

  private Environment env;

  @Inject public LoginController(FormFactory f, Environment theEnv) {
    super(f);
    env = theEnv;
  }

  public Result authenticate() {
    if(env.isProd())
      // Disable login in prod mode
      return redirect(routes.LoginController.login());

    DynamicForm form = factory().form().bindFromRequest();
    String userName = form.get(StringConsts$.MODULE$.NAME_NAME());
    String passwort = form.get(StringConsts$.MODULE$.PW_NAME());

    User user = findOrCreateStudent(userName, passwort);

    session().clear();
    session(StringConsts$.MODULE$.SESSION_ID_FIELD(), user.name);

    return redirect(controllers.routes.Application.index());
  }

  public Result fromWuecampus(String userName, int courseId, String courseName) {
    if(userName.isEmpty())
      return redirect(routes.LoginController.login());

    String passwort = "";
    User student = findOrCreateStudent(userName, passwort);

    if(Course$.MODULE$.finder().byId(courseId) == null && courseId != -1) {
      // Create course
      Course course = new Course();
      course.name_$eq(courseName);
      course.save();
    }

    session().clear();
    session(StringConsts$.MODULE$.SESSION_ID_FIELD(), student.name);

    return redirect(controllers.routes.Application.index());
  }

  public Result login() {
    return ok(views.html.login.render());
  }

  public Result logout() {
    session().clear();
    return redirect(routes.LoginController.login());
  }

  private User findOrCreateStudent(String userName, String passwort) {
    // TODO: Passwort!
    User user = User.finder.byId(userName);

    if(user != null)
      return user;

    user = new User();
    user.name = userName;
    user.save();

    // FIXME: get toolObject??
    //    Path solutionDirectory = getSolDirForUser(userName);
    //    if(solutionDirectory.toFile().exists())
    //      return user;
    //
    //    try {
    //      Files.createDirectories(solutionDirectory);
    //    } catch (IOException e) {
    //      Logger.error("Could not create solution directory for user " + userName, e);
    //    }

    return user;
  }

}

File: app/controllers/LoginController.scala
package controllers

import com.github.t3hnar.bcrypt._
import javax.inject._
import model.core.CoreConsts._
import model.core.Repository
import model._
import play.api.data.Form
import play.api.db.slick.{DatabaseConfigProvider, HasDatabaseConfigProvider}
import play.api.mvc._
import slick.jdbc.JdbcProfile

import scala.concurrent.{ExecutionContext, Future}

class LoginController @Inject()(cc: ControllerComponents, val dbConfigProvider: DatabaseConfigProvider, val repo: Repository)(implicit ec: ExecutionContext)
  extends AbstractController(cc) with HasDatabaseConfigProvider[JdbcProfile] with play.api.i18n.I18nSupport {

  def registerForm: Action[AnyContent] = Action {
    implicit request => Ok(views.html.registerForm(FormMappings.userCredForm))
  }

  def register: Action[AnyContent] = Action.async { implicit request =>

    val onError: Form[UserCredentials] => Future[Result] = { _ =>
      Future(BadRequest("There has been an error in your form..."))
    }

    val onRead: UserCredentials => Future[Result] = { credentials =>
      val newUser = RegisteredUser(credentials.username)
      val pwHash = PwHash(credentials.username, credentials.password.bcrypt)

      repo.saveUser(newUser).flatMap {
        case false => Future(BadRequest("Could not save user!"))
        case true  =>
          repo.savePwHash(pwHash) map {
            _ => Ok(views.html.registered(credentials.username))
          }
      }
    }

    FormMappings.userCredForm.bindFromRequest.fold(onError, onRead)
  }

  def authenticate: Action[AnyContent] = Action.async { implicit request =>

    val onError: Form[UserCredentials] => Future[Result] = { formWithErrors =>
      Future(BadRequest(views.html.loginForm(formWithErrors)))
    }

    val onRead: UserCredentials => Future[Result] = { credentials =>

      val futureUserAndPwHash: Future[(Option[User], Option[PwHash])] = for {
        user <- repo.userByName(credentials.username)
        pwHash <- repo.pwHashForUser(credentials.username)
      } yield (user, pwHash)

      futureUserAndPwHash map {
        case (None, _)                  => Redirect(controllers.routes.LoginController.register())
        case (Some(_), None)            => BadRequest("Cannot change password!")
        case (Some(user), Some(pwHash)) =>
          if (credentials.password isBcrypted pwHash.pwHash) {
            Redirect(controllers.routes.Application.index()).withSession(sessionIdField -> user.username)
          } else {
            Ok(views.html.loginForm(FormMappings.userCredForm.fill(credentials)))
          }
      }
    }

    FormMappings.userCredForm.bindFromRequest.fold(onError, onRead)
  }

  def loginForm: Action[AnyContent] = Action {
    implicit request => Ok(views.html.loginForm(FormMappings.userCredForm))
  }

  def logout: Action[AnyContent] = Action {
    implicit request => Redirect(routes.LoginController.loginForm()).withNewSession
  }

}
